### PR TITLE
Support small FlexAttention score_mod reductions

### LINKED ISSUE #152593: Flex Attention doesn't scale with custom bias
### 🐛 Describe the bug

When using FlexAttention with the following custom `RBFBias`, I cannot get FlexAttention to go above ~L=1600 without OOMing. I also find that this implementation is about 6-8x slower than an implementation in JAX. Is there something I can configure to leverage the memory efficiency of other fused attention mechanisms?

```python
#!/usr/bin/env -S PYENV_VERSION=torch python3
import argparse
import sys
import time

import torch
import torch.nn as nn
from torch.nn.attention.flex_attention import flex_attention

torch.set_default_device("cuda:0")


def main(seed: int, N: int, B: int, H: int, L: int, D: int, D_s: int):
    torch.manual_seed(seed)
    bias = RBFBias(num_heads=4, num_basis=5)
    kernel_options = {}
    # kernel_options = { # NOTE: doesn't help
    #     "BLOCK_M": 64,
    #     "BLOCK_N": 64,
    #     "BLOCK_M1": 32,
    #     "BLOCK_N1": 64,
    #     "BLOCK_M2": 64,
    #     "BLOCK_N2": 32,
    # }  # https://github.com/pytorch/pytorch/issues/133254
    attn = BiasedFlexAttention(bias, kernel_options)
    # attn = torch.compile(attn, dynamic=False)  # NOTE: fails
    b = sample_batch(B, H, L, D, D_s)
    attn(**b)  # precompile flex_attention (?)
    torch.cuda.synchronize()
    times = torch.zeros(N)
    for i in range(N):
        b = sample_batch(B, H, L, D, D_s)
        torch.cuda.synchronize()
        start = time.perf_counter()
        attn(**b)
        torch.cuda.synchronize()
        stop = time.perf_counter()
        times[i] = stop - start
    print(
        f"[B={B} H={H} L={L} D={D} D_s={D_s}]: {times.mean():0.6f}±{times.std():0.6f}s"
    )


class RBFBias(nn.Module):
    def __init__(self, num_heads, num_basis):
        super().__init__()
        self.alpha = nn.Parameter(torch.randn(num_heads, num_basis))
        self.beta = nn.Parameter(torch.randn(num_heads, num_basis))

    def forward(self, score, b, h, q_idx, kv_idx, qs_s, ks_s):
        q_s = qs_s[b, q_idx]
        k_s = ks_s[b, kv_idx]
        d_sq = torch.square(q_s - k_s).sum()
        alpha, beta = self.alpha[h], self.beta[h]
        d_rbf = alpha * torch.exp(-beta * d_sq)
        return score + d_rbf.sum()


class BiasedFlexAttention(nn.Module):
    def __init__(self, bias: nn.Module, kernel_options: dict = {}):
        super().__init__()
        self.bias = bias
        self.kernel_options = kernel_options

    def forward(self, qs, ks, vs, qs_s, ks_s):
        def score_mod(score, b, h, q_idx, kv_idx):
            return self.bias(score, b, h, q_idx, kv_idx, qs_s, ks_s)

        return flex_attention(
            qs,
            ks,
            vs,
            score_mod=score_mod,
            kernel_options=self.kernel_options,
        )


def sample_batch(B: int, H: int, L: int, D: int, D_s: int):
    qs, ks, vs = torch.randn(3, B, H, L, D)
    qs_s, ks_s = torch.randn(2, B, L, D_s)
    return {"qs": qs, "ks": ks, "vs": vs, "qs_s": qs_s, "ks_s": ks_s}


def parse_args(argv):
    parser = argparse.ArgumentParser(
        prog=argv[0],
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
    )
    parser.add_argument(
        "-seed",
        type=int,
        default=42,
        help="Seed.",
    )
    parser.add_argument(
        "-N",
        type=int,
        default=1000,
        help="Number of trials.",
    )
    parser.add_argument(
        "-B",
        type=int,
        default=32,
        help="Batch size.",
    )
    parser.add_argument(
        "-L",
        type=int,
        default=1664,  # 1024 + 512 + 128
        help="Length of sequence to test.",
    )
    parser.add_argument(
        "-H",
        type=int,
        default=4,
        help="Number of attention heads.",
    )
    parser.add_argument(
        "-D",
        type=int,
        default=16,
        help="Embedding dim per head.",
    )
    parser.add_argument(
        "-D_s",
        type=int,
        default=2,
        help="Spatial dimension.",
    )
    return parser.parse_args(argv[1:])


if __name__ == "__main__":
    args = parse_args(sys.argv)
    main(**vars(args))
```

### Versions

```
PyTorch version: 2.7.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Void Linux (x86_64)
GCC version: (GCC) 14.2.1 20250405
Clang version: 17.0.6
CMake version: version 3.30.1
Libc version: glibc-2.41

Python version: 3.12.7 (main, Oct  6 2024, 23:32:45) [GCC 13.2.0] (64-bit runtime)
Python platform: Linux-6.12.25_1-x86_64-with-glibc2.41
Is CUDA available: True
CUDA runtime version: 12.6.85
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4090
Nvidia driver version: 570.144
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        48 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               32
On-line CPU(s) list:                  0-31
Vendor ID:                            AuthenticAMD
Model name:                           AMD Ryzen 9 7950X 16-Core Processor
CPU family:                           25
Model:                                97
Thread(s) per core:                   2
Core(s) per socket:                   16
Socket(s):                            1
Stepping:                             2
Frequency boost:                      enabled
CPU(s) scaling MHz:                   35%
CPU max MHz:                          5881.0000
CPU min MHz:                          545.0000
BogoMIPS:                             8982.57
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl xtopology nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local user_shstk avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid overflow_recov succor smca fsrm flush_l1d amd_lbr_pmc_freeze
Virtualization:                       AMD-V
L1d cache:                            512 KiB (16 instances)
L1i cache:                            512 KiB (16 instances)
L2 cache:                             16 MiB (16 instances)
L3 cache:                             64 MiB (2 instances)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-31
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Mitigation; Safe RET
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==2.2.3
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] pytorch-lightning==2.5.1
[pip3] torch==2.7.0
[pip3] torchmetrics==1.7.0
[pip3] torchvision==0.21.0
[pip3] triton==3.3.0
[conda] Could not collect
```

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185611

FlexAttention lowers score_mod as a pointwise subgraph that is inlined into the surrounding template, so that lowering path cannot create temporary buffers. Small reductions are already converted into inline Pointwise IR, but the RBF-bias pattern broadcasts one of those reduced scalar expressions into a basis-vector expression. The generic reuse heuristic then tried to realize the inline expression as a ComputedBuffer, and PointwiseSubgraphLowering rejected the buffer allocation.

For trainable RBF bias parameters, backward hit a second related limitation. The captured-gradient scatter for alpha[h, :] and beta[h, :] produces a small vector slice per score_mod invocation, but the FlexAttention backward template only emitted scalar captured-gradient stores.

Fix this by suppressing heuristic mark_reuse realization while lowering pointwise subgraphs, while leaving explicit realize paths intact so large non-unrolled reductions still fail normally. Also teach the FlexAttention backward template to unroll small static captured-gradient scatter slices, and reject larger slices before template selection using the same strict-below unroll threshold policy as reductions.

Fixes #152593
Generated by my agent

Benchmark Results:
Before the fix, the RBF score_mod benchmark failed during compilation with `SubgraphLoweringException: Buffers cannot be created while lowering a pointwise subgraph`, so no steady-state timing was available.

After the fix, a trainable RBF fwd+bwd smoke benchmark on the local NVIDIA B200 with `B=2 H=2 L=256 D=32 D_s=2 basis=5` produced:

```
warmup_compile_plus_first_s=5.026798
iter_s=0.003430,0.002700,0.002532,0.002438,0.002419
mean_s=0.002704
peak_allocated_mib=1.02
grad_means=0.000001,0.000001,0.000003,0.000093,0.000057
```

Test Plan:
```
python test/inductor/test_flex_attention.py -k test_score_mod_rbf_bias_trainable_reductions -v
python test/inductor/test_flex_attention.py -k test_score_mod_captured_grad_slice_unroll_limit -v
python test/inductor/test_flex_attention.py -k test_captured_scalar_grad -v
python test/inductor/test_flex_attention.py -k test_reduction_unrolled -v
python test/inductor/test_flex_attention.py -k test_cant_lower_error_message -v
python test/inductor/test_flex_attention.py -k test_captured_wrong_device_error_message -v
python test/inductor/test_flex_attention.py -k test_bf16_score_mod_captured_grad_dtype -v
lintrunner -a
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by drisspg — CHANGES_REQUESTED** [review]
This is a non trivial change to the semantics of flash that I would like to grok further and not have an agent handle
